[upd] this expressions (#5657)
JetBrains/kotlin-web-site: docs/topics/this-expressions.md

Reviewer: Since we're here, let's make this runnable too? :)

Author: But it's already runnable?

Reviewer: GitHub doesn't want to show me the last line apparently. Probably because we need an empty line before EOF.

[//]: # (title: This expressions)

> The `this` expression refers to the current receiver. How you use `this` depends on the context:

Reviewer: It's a little off-topic, but to someone new to Kotlin it's not actually clear what a receiver is, whether we mark it as a term or not. Since it's a relatively general OOP term, we can probably link somewhere to clarify, but I bet as is a lot of people either stumble on this or skip forward.

Author: I wanted to do so, but I'm not sure there is a good ebough place to link to tbh. The wiki page is okay, but I don't like that it's called command pattern

Reviewer: We can just define it right here, "receiver (object in which context the code is executed)" or something. The wiki page is highly abstract and linking there I don't think will add a lot of value.

> * In a member of a [class](classes.md#inheritance), `this` refers to the current object of that class.

Reviewer: ```suggestion
* In a member of a [class](classes.md), `this` refers to the current object of that class.
```

  For example, in the following code, `this.name` means the `name` property of the current `Language` object:

  ```kotlin
  //sampleStart
  class Language(val name: String) {
      fun printName() {
          println(this.name)
      }
  }

  fun main() {
      val language = Language("Kotlin")
      language.printName()
      // Kotlin
  }
  //sampleEnd
  ```
  {kotlin-runnable="true"}

> * In an [extension function](extensions.md) or a [function literal with receiver](lambdas.md#function-literals-with-receiver),

Reviewer: Links forward (in the context of the ToC at least) are always risky, and actually I wonder whether `this` should be under "Functions" at all. It feels like it belongs both to "Classes" and "Functions" which usually means it belongs to neither.

Author: Agree, I tried different places but I don't want to be the one who suggests to change the toc AGAIN :D

`this` refers to the receiver. 

>   In the following example, we call `lastCharacter()`on the string

Reviewer: ```suggestion
  In the following example, `lastCharacter()` is called on the string
```

To avoid first person where it's not necessary.

Author: It's funny cause I used first person to avoid passive :D

  `"Kotlin"`. Therefore, `"Kotlin"` is the receiver. Inside the extension function, `this` refers
  to `"Kotlin"`. This means that `this.length` is the length of `"Kotlin"`:

  ```kotlin
  //sampleStart
  fun String.lastCharacter(): Char {
      println(this.length)
      // 6
      return this[this.length - 1]
  }

  fun main() {
      println("Kotlin".lastCharacter())
      // n
  }
  //sampleEnd
  ```
  {kotlin-runnable="true"}

> > In simple cases, you don't need to write `this` explicitly. Kotlin interprets it from the current scope.

Reviewer: ```suggestion
> In simple cases, you don't need to write `this` explicitly. Kotlin resolves it from the current scope.
```

Nitnit.

> > Learn more about [](#implicit-this).

Reviewer: Should we simply put the "Implicit `this`" section next instead of linking to the end of the same page? It's small enough that it shouldn't affect the flow in the same way as going to the bottom and then returning back up.

Author: It would make sense, but it describes qualified vs unqualified calls that we explain later

Reviewer: Actually I think I misunderstood here, it's fine as is :)

> 
{style="tip"}

> ## Qualified this

Reviewer: ```suggestion
## Qualified `this`
```

            println(this.value)
            // 22
            
            // Refers to the name property of the outer User object 
            println(this@User.name)
            // Jane Doe
        }
    }
}

fun main() {
    val user = User("Jane Doe")
    val age = user.Age(22)
    age.printInfo()
}
//sampleEnd
```
{kotlin-runnable="true"}

> Only inner classes hold a reference to an outer class instance. Regular [nested classes](nested-classes.md) don't have access to
> `this` from the outer class.
> 
{style="note"}

### Access a class from an extension function

If you declare an extension function inside a class, two receivers are available:

* The extension receiver: the value on which you call the extension function.
* The dispatch receiver: the current object of the class where you declare the extension function.

Use qualified `this` to specify which receiver you need:

```kotlin
//sampleStart
class User {
    val prefix = "Name"
    
    fun String.formatName(): String {
        return "${this@User.prefix}: ${this.uppercase()}"
    }

    fun printUser() {
        println("Jane Doe".formatName())
    }
}

fun main() {
    val user = User()
    user.printUser()

>     // Name: JANE DOE

Reviewer: Nit: a `User` class that doesn't have even a hardcoded name property reads very synthetic to me :) it's an extra line, but still.

}
//sampleEnd
```
{kotlin-runnable="true"}

Here:

> * `this` refers to the extension receiver (the `"Jane Doe"` string).

Reviewer: ```suggestion
* `this.uppercase()` refers to the extension function of `String`.
```

Reviewer: Simply `this` is hard to attribute.

Author: Agree, but `uppercase()` isn’t an extension function of String in the sense your wording explains. More importantly, the point is that `this` is the extension receiver, not that `uppercase()` is an extension. I have rewritten that bullets, let me know what of think of them!

* `this@User` refers to the current `User` object.

> * `this@User.prefix` accesses the prefix property of the current `User` object.

Reviewer: ```suggestion
* `this@User.prefix` accesses the `prefix` property of the current `User` object.
```

### Access from a lambda

> Unlike a regular lambda, a [lambda with receiver](lambdas.md#function-literals-with-receiver) has its own receiver. Therefore, `this` inside it

Reviewer: "Lambda with receiver has a receiver" is funny :)

Reviewer: There is a term conflict here: above the very same link has text "function literal with receiver", here it's "lambda with receiver". We should the same term to refer to the same thing; probably "lambda" since it's the shorter expression (and that's why we use in the rest of the docs)?

Author: Hm, but the page to whichj we link says: lambda expressions can be used as function literals with receiver when the receiver type can be inferred from the context. So I wouldn't treat these terms as identical.

> Unlike a regular lambda, a [lambda with receiver](lambdas.md#function-literals-with-receiver) has its own receiver. Therefore, `this` inside it

Reviewer: "`this` inside it refers to that receiver" is not illustrated in the sample, and I think it would be beneficial to see the examples side-by-side like in other sections.

> refers to that receiver. If a lambda with receiver is nested inside another receiver scope, add a label to the lambda

Reviewer: I think the key to the funny is "that receiver" which is not immediately opposed by something else. "To that receiver... instead of an outside scope".

and use qualified `this` to specify which receiver you need:

```kotlin
//sampleStart
class User(val name: String) {
    fun printWithPrefix() {

>         val print: String.() -> Unit = stringLabel@ {

Reviewer: Talking about name resolution, maybe let's not call custom stuff after stdlib functions :)
```suggestion
        val printString: String.() -> Unit = stringLabel@ {
```

Author: Oh, good point! Thanks

            println("${this@stringLabel}: ${this@User.name}")
        }
      

>         print("User")

Reviewer: ```suggestion
        printString("User")
```
See above.

    }
}

fun main() {
    val user = User("Jane Doe")
    user.printWithPrefix()
    // User: Jane Doe
}
//sampleEnd
```
{kotlin-runnable="true"}

Here: 

* `stringLabel@` labels the lambda.
* `this@stringLabel` refers to the string receiver of the lambda.
* `this@User` refers to the current `User` object.

The label doesn't call anything or change how the lambda works. It only helps you refer to the lambda's receiver.

### Access an anonymous object

> Since anonymous objects don't have class names, you can't use a name like `this@YourClassName` to refer

Reviewer: ```suggestion
Since [anonymous objects](object-declarations.md#object-expressions) don't have class names, you can't use a name like `this@YourClassName` to refer
```

> to it. Inside an anonymous object, use plain `this` to refer to the object:

Reviewer: ```suggestion
to it. Inside an anonymous object, use unqualified `this` to refer to the object:
```

```kotlin
//sampleStart
interface UserPrinter {
    fun print()
}

fun printUser(printer: UserPrinter) {
    printer.print()
}

fun main() {
    val printer = object : UserPrinter {
        val name = "Jane Doe"
      
        override fun print() {
            // Refers to the name property of the anonymous object
            println(this.name)
        }
    }
    printUser(printer)
    // Jane Doe
}

> //sampleEnd

Reviewer: Since `// sampleStart` and `// sampleEnd` coincide with the code block, can we get rid of them? The rest of the samples are the same, maybe I'm missing something.

```
{kotlin-runnable="true"}

> If you create an anonymous object inside another class, you can use qualified `this` to access the outer class object.

> > Use it to access an outer receiver, not the anonymous object itself.

Reviewer: I think I'd appreciate a sample for this, too, as it's hardly a side note. And the section may be called "Access an anonymous object or its outer class", since we don't have that covered anywhere else on the page — even though it's arguably intuitive from the basic rules of `this`.

Author: done

> 
{style="tip"}

> ## Implicit this

Reviewer: ```suggestion
## Implicit `this`
```

[//]: # (title: This expressions)

> The `this` expression refers to the current object or receiver.

Reviewer 2: To be honest, I'm not a fan of saying "current object or receiver". `this` always refers to the current receiver, and it could be an instance of a class or interface, or an object.

The `this` expression refers to the current object or receiver.

How you use `this` depends on the context:

* In a member of a [class](classes.md#inheritance), `this` refers to the current object of that class.
* In an [extension function](extensions.md) or a [function literal with receiver](lambdas.md#function-literals-with-receiver)
`this` denotes the _receiver_ parameter that is passed on the left-hand side of a dot.

> If `this` has no qualifiers, it refers to the _innermost enclosing scope_. To refer to `this` in other scopes, _label qualifiers_ are used:

Reviewer 2: This is actually very important: `this` doesn't only refer to the current receiver. It actually can refer to any of the nested receivers. For example, if I have

```kotlin
class Foo {
  fun foo() { }
}

class Bar {
  fun bar() { }

  fun Foo.moo() {
    this.bar()  // 'this' refers to the enclosing 'Bar' instance
    this.foo()  // 'this' refers to the extension receiver
  }
}
```

Author: Added to the Qualified this section. Do you think it also makes sense to include the example like the one you've provided?

Reviewer 2: I still think this is incorrect. You write "If `this` has no qualifier, it refers to the receiver of the _innermost enclosing scope_", but my example above shows that `this` may refer to different enclosing scopes, even if it appears unqualified.

Author: Got it now, thank you! What do you think about smth like this:

Kotlin code can have several receivers available at the same time. This happens when receiver scopes are nested. For example, an extension function declared inside a class can access members of both the extension receiver and the enclosing class object.
Kotlin can use these receivers implicitly when you access their members. Unqualified this refers to the receiver with the highest priority in the current scope. To refer to another receiver explicitly, use a qualified this.

Reviewer 2: The problem with your comment is in the "Unqualified this refers to the receiver with the highest priority in the current scope." part, unqualified `this` may refer to any receiver in scope, the closest one is just given more priority. So if, for example, several receivers have the same type, using the qualified `this` is the only possibility to access the outer one.

Author: Yeah in that case my wording is too restrictive ig, so is the following would be correct: 

Your code can have several receivers available at the same time when receiver scopes are nested. Kotlin can use any available receiver to access its members implicitly, but receivers from inner scopes have higher priority. To explicitly refer to a particular receiver, use a qualified this. This is especially useful when multiple receivers have members with the same name and you need to access a member of an outer receiver.

Reviewer 2: That's much better, thanks!

  class Language(val name: String) {
      fun printName() {
          println(this.name)
      }
  }

  fun main() {
      val language = Language("Kotlin")
      language.printName()
      // Kotlin
  }
  //sampleEnd
  ```
  {kotlin-runnable="true"}

* In an [extension function](extensions.md) or a [function literal with receiver](lambdas.md#function-literals-with-receiver),
`this` refers to the receiver. 

  A _receiver_ is the object that the function works with. In the following example, we call `lastCharacter()`on the string
  `"Kotlin"`. Therefore, `"Kotlin"` is the receiver. Inside the extension function, `this` refers
  to `"Kotlin"`. This means that `this.length` is the length of `"Kotlin"`:

  ```kotlin
  //sampleStart
  fun String.lastCharacter(): Char {
      println(this.length)
      // 6
      return this[this.length - 1]
  }

  fun main() {
      println("Kotlin".lastCharacter())
      // n
  }
  //sampleEnd
  ```
  {kotlin-runnable="true"}

> > In simple cases, you don't need to write `this` explicitly. Kotlin infers it from the current scope.

Reviewer 2: The use of the word _infers_ is a bit overloaded in Kotlin already, and mostly refers to type inference.

    fun printWithPrefix() {
        val printString: String.() -> Unit = stringLabel@ {
            println("${this@stringLabel}: ${this@User.name}")
        }

        printString("User")
    }
}

fun main() {
    val user = User("Jane Doe")
    user.printWithPrefix()
    // User: Jane Doe
}
```
{kotlin-runnable="true"}

Here: 

* `stringLabel@` labels the lambda.
* `this@stringLabel` refers to the string receiver of the lambda.
* `this@User` refers to the current `User` object.

The label doesn't call anything or change how the lambda works. It only helps you refer to the lambda's receiver.

### Access an anonymous object or its outer class

An [anonymous object](object-declarations.md#object-expressions) has its own receiver scope. Inside the object body, unqualified `this` refers to the anonymous object
itself. However, since anonymous objects don't have class names, you can't use them as `this` qualifiers. Therefore,
you can refer to the anonymous object only with unqualified `this`:

```kotlin
interface UserPrinter {
    fun print()
}

fun main() {
    val printer = object : UserPrinter {
        val prefix = "User"
        
        override fun print() {

>             // this refers to the anonymous object, so this.prefix accesses its `prefix` property

Reviewer: ```suggestion
            // `this` refers to the anonymous object, so `this.prefix` accesses its `prefix` property
```
With the human-readable syntax it's hard to see immediately whether the lower-case letter is a typo or a keyword (it's all the more confusing because `prefix` is marked). I don't think we have a guide line for that, but it just makes sense™

            println(this.prefix)
        }
    }
    printer.print()
    // User
}
```
{kotlin-runnable="true"}

If you declare an anonymous object inside another class, you can use qualified `this` to access the enclosing object:

```kotlin

interface UserPrinter {
    fun print()
}

class User(val name: String) {
    fun createPrinter(): UserPrinter {
        return object : UserPrinter {
            override fun print() {

>                 // this@User refers to the enclosing `User` object, so this@User.name accesses its name property

Reviewer: I can just see the warning about the code line being too long for the web page, let's add a newline after the comma?
